pkg/controller/secretsyncer/secretsyncer.go

```diff
@@ -0,0 +1,231 @@
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get secret %s: %w", sourceSecretName.String(), err)
+	}
+	if source.DeletionTimestamp != nil {
+		log.Debug("not doing work for secret because it is being deleted")
+		return nil
+	}
+
+	if len(source.Data) == 0 {
+		log.Trace("not updating target secret as source has no data")
+		return nil
+	}
+
+	var mirrorErrors []error
+	for _, mirrorConfig := range r.config().Secrets {
+		if mirrorConfig.From.Namespace != sourceSecretName.Namespace || mirrorConfig.From.Name != sourceSecretName.Name {
+			continue
+		}
+
+		// Verify this doesn't target anything managed by ci-secret-boostrap
+		if allowed := r.targetFilter(cluster, types.NamespacedName{Namespace: mirrorConfig.To.Namespace, Name: mirrorConfig.To.Name}); !allowed {
+			// Log a big fat error, but don't return it as that would result in retrying.
+			log.WithField("target", mirrorConfig.String()).Error("Ignoring secret, its target is managed by ci-secret-boostrap")
+			continue
+		}
+		// TODO: Ideally we would would create one reconcile.Request per target. This here means
+		// that if a single target is broken, it will send all targets into exponential backoff.
+		mirrorErrors = append(mirrorErrors, r.mirrorSecret(log, client, source, mirrorConfig.To))
+	}
+
+	return utilerrors.NewAggregate(mirrorErrors)
+}
+
+func (r *reconciler) mirrorSecret(log *logrus.Entry, c ctrlruntimeclient.Client, source *corev1.Secret, to config.SecretLocation) error {
+	*log = *log.WithFields(logrus.Fields{"target-namespace": to.Namespace, "target-secret": to.Name})
+
+	secret, mutateFn := secret(
+		types.NamespacedName{Namespace: to.Namespace, Name: to.Name},
+		source.Data,
+		source.Type,
+	)
+	result, err := crcontrollerutil.CreateOrUpdate(r.ctx, c, secret, mutateFn)
+	if err != nil {
+		// Happens if someone changes the type, just re-create it
+		if strings.Contains(err.Error(), "field is immutable") {
+			if err := recreateSecret(r.ctx, c, secret); err != nil {
+				return fmt.Errorf("failed to recreate secret %s/%s: %w", secret.Namespace, secret.Name, err)
+			}
+		}
+		return fmt.Errorf("failed to upsert secret: %w", err)
+	}
+	if result != crcontrollerutil.OperationResultNone {
+		log.WithField("operation", result).Info("Upsert succeeded")
+	}
+	return nil
+}
+
+func recreateSecret(ctx context.Context, c ctrlruntimeclient.Client, s *corev1.Secret) error {
+	if err := c.Delete(ctx, s.DeepCopy()); err != nil {
+		return fmt.Errorf("failed to delete secret: %w", err)
+	}
+	if err := c.Create(ctx, s); err != nil {
+		return fmt.Errorf("failed to create secret: %w", err)
+	}
+	return nil
+}
+
+func secret(nn types.NamespacedName, data map[string][]byte, tp corev1.SecretType) (*corev1.Secret, crcontrollerutil.MutateFn) {
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: nn.Namespace,
+			Name:      nn.Name,
+		},
+	}
+	return s, func() error {
```

Comment:
The label indicates it is generated by `ci-secret-bootstrap`:

https://github.com/openshift/ci-tools/blob/1a991c8c1437a3387bcd2cdd04f58d535b5f9715/cmd/ci-secret-bootstrap/main.go#L327

Maybe for this tool:

`Labels:    map[string]string{"ci.openshift.org/auto-managed-by-controller": "true"}, `

Reply:
Done

Reply:
LGTM.

@stevekuznetsov WDYT?